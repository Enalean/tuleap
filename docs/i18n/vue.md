# i18n in Vue apps

We use Gettext with
[vue-gettext](https://github.com/Polyconseil/vue-gettext) to translate
strings in our Vue apps. Here is some advice specific to `vue-gettext`:

## Always add a `key` attribute when you have conditional translated strings

For example, DO NOT do this:

``` html
// TranslatedExample.vue
<template>
    <p v-translate v-if="a_condition">First translated string</p>
    <p v-translate v-else>This string will never be shown</p> <!-- <== This string will NEVER be shown -->
</template>
```

The string in the `v-else` will never be shown, because Vue does not
know that the two `<p>` tags actually have different content. Vue tries
to limit the number of DOM changes, so it will only change attributes
and will change the text node, which will mess up `vue-gettext`. See
[the Vue.js documentation on the key
attribute](https://v2.vuejs.org/v2/api/#key) for details.

INSTEAD, DO THIS:

``` html
// TranslatedExample.vue
<template>
    <p v-translate v-if="a_condition" key="first_case">First translated string</p>
    <p v-translate v-else key="other_case">This string will be shown</p>
</template>
```

## Never use Vue.js interpolation inside translated strings

For example, DO NOT do this:

``` html
// TranslatedExample.vue
<template>
    <p v-translate>Current value: {{ reactive_value }}</p>
    <translate>Current value: {{ reactive_value }}</translate>
</template>
```

This will break the translation. The string will always show in English,
never in translated languages. Always use
`v-bind:translate-params="{ params }"` or `v-translate="{ params }"`
with `%{ param }` in the translated string. See [vue-gettext\'s
syntax](https://github.com/Polyconseil/vue-gettext#custom-parameters)

INSTEAD, DO THIS:

``` html
// TranslatedExample.vue
<template>
   <p v-translate="{ reactive_value }">Current value: %{ reactive_value }</p>
   <translate v-bind:translate-params="{ reactive_value }">Current value: %{ reactive_value }</translate>
</template>
```

## Never use `v-bind` on attributes in HTML tags in translated strings

For example, DO NOT do this:

``` html
// TranslatedExample.vue
<template>
    <p v-translate><a v-bind:href="link_url">{{ link_text }}</a> has done some changes in this document.</p>
</template>
```

This will break reactivity. If `link_url` or `link_text` change value,
the text will not change. See [vue-gettext\'s doc about
this](https://github.com/Polyconseil/vue-gettext#caveat-when-using-v-translate-with-vue-components-or-vue-specific-attributes).

INSTEAD, DO THIS:

``` html
// TranslatedExample.vue
<template>
    <p v-translate="{ link_url, link_text }"><a href="%{ link_url }">%{ link_text }</a> has done some changes in this document.</p>
</template>
```

## Name your parameter when your translations have parameters

For example, DO NOT do this:

``` html
// TranslatedExample.vue
<template>
    <p v-bind:translate-params="vue_variable_for_nb">%{ vue_variable_for_nb } changes have been done in this document.</p>
</template>
```

If your vue variable is updated, then you won\'t have to update the
corresponding translation.

INSTEAD, DO THIS:

``` html
// TranslatedExample.vue
<template>
    <p v-bind:translate-params="{nb: vue_variable_for_nb}">%{ nb } changes have been done in this document.</p>
</template>
```

## Resources

-   vue-gettext: <https://github.com/Polyconseil/vue-gettext>
