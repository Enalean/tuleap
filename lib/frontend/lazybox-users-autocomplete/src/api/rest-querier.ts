/**
 * Copyright (c) Enalean, 2024 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { User } from "@tuleap/core-rest-api-types";
import type { ResultAsync } from "neverthrow";
import type { Fault } from "@tuleap/fault";
import { uri, getJSON } from "@tuleap/fetch-result";

export const fetchMatchingUsers = (query: string): ResultAsync<User[], Fault> => {
    return getJSON(uri`/api/v1/users`, {
        params: {
            query,
            limit: 10,
            offset: 0,
        },
    });
};
