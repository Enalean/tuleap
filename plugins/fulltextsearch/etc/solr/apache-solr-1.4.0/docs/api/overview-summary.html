<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_15) on Fri Nov 06 12:35:41 EST 2009 -->
<TITLE>
Overview (Solr 1.4.0 API)
</TITLE>

<META NAME="date" CONTENT="2009-11-06">

<LINK REL ="stylesheet" TYPE="text/css" HREF="stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Overview (Solr 1.4.0 API)";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Overview</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?overview-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="overview-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H1>
Solr 1.4.0 API (1.4.0)
</H1>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Core</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/analysis/package-summary.html">org.apache.solr.analysis</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/common/package-summary.html">org.apache.solr.common</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/core/package-summary.html">org.apache.solr.core</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/package-summary.html">org.apache.solr.handler</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/admin/package-summary.html">org.apache.solr.handler.admin</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/component/package-summary.html">org.apache.solr.handler.component</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/highlight/package-summary.html">org.apache.solr.highlight</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/request/package-summary.html">org.apache.solr.request</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/schema/package-summary.html">org.apache.solr.schema</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/search/package-summary.html">org.apache.solr.search</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/search/function/package-summary.html">org.apache.solr.search.function</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/servlet/package-summary.html">org.apache.solr.servlet</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/servlet/cache/package-summary.html">org.apache.solr.servlet.cache</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/spelling/package-summary.html">org.apache.solr.spelling</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/tst/package-summary.html">org.apache.solr.tst</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/update/package-summary.html">org.apache.solr.update</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/update/processor/package-summary.html">org.apache.solr.update.processor</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/util/package-summary.html">org.apache.solr.util</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/util/plugin/package-summary.html">org.apache.solr.util.plugin</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/util/xslt/package-summary.html">org.apache.solr.util.xslt</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Common</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/common/luke/package-summary.html">org.apache.solr.common.luke</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/common/params/package-summary.html">org.apache.solr.common.params</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/common/util/package-summary.html">org.apache.solr.common.util</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>SolrJ</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/package-summary.html">org.apache.solr.client.solrj</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/beans/package-summary.html">org.apache.solr.client.solrj.beans</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/embedded/package-summary.html">org.apache.solr.client.solrj.embedded</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/impl/package-summary.html">org.apache.solr.client.solrj.impl</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/request/package-summary.html">org.apache.solr.client.solrj.request</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/response/package-summary.html">org.apache.solr.client.solrj.response</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/client/solrj/util/package-summary.html">org.apache.solr.client.solrj.util</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>contrib: DataImportHandler</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/dataimport/package-summary.html">org.apache.solr.handler.dataimport</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>contrib: Clustering</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/clustering/package-summary.html">org.apache.solr.handler.clustering</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/clustering/carrot2/package-summary.html">org.apache.solr.handler.clustering.carrot2</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>contrib: Solr Cell</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/apache/solr/handler/extraction/package-summary.html">org.apache.solr.handler.extraction</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Overview</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?overview-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="overview-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>
Copyright &copy; 2009 Apache Software Foundation.  All Rights Reserved.
</BODY>
</HTML>
