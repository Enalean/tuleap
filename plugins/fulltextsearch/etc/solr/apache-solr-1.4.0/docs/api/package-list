org.apache.solr.analysis
org.apache.solr.client.solrj
org.apache.solr.client.solrj.beans
org.apache.solr.client.solrj.embedded
org.apache.solr.client.solrj.impl
org.apache.solr.client.solrj.request
org.apache.solr.client.solrj.response
org.apache.solr.client.solrj.util
org.apache.solr.common
org.apache.solr.common.luke
org.apache.solr.common.params
org.apache.solr.common.util
org.apache.solr.core
org.apache.solr.handler
org.apache.solr.handler.admin
org.apache.solr.handler.clustering
org.apache.solr.handler.clustering.carrot2
org.apache.solr.handler.component
org.apache.solr.handler.dataimport
org.apache.solr.handler.extraction
org.apache.solr.highlight
org.apache.solr.request
org.apache.solr.schema
org.apache.solr.search
org.apache.solr.search.function
org.apache.solr.servlet
org.apache.solr.servlet.cache
org.apache.solr.spelling
org.apache.solr.tst
org.apache.solr.update
org.apache.solr.update.processor
org.apache.solr.util
org.apache.solr.util.plugin
org.apache.solr.util.xslt
