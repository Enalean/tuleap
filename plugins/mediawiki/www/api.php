/usr/share/mediawiki-tuleap/api.php