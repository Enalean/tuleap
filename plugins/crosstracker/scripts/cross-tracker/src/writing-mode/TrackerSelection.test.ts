/*
 * Copyright (c) Enalean, 2018-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { Mock } from "vitest";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { nextTick } from "vue";
import type { VueWrapper } from "@vue/test-utils";
import { shallowMount } from "@vue/test-utils";
import { errAsync, okAsync } from "neverthrow";
import { Fault } from "@tuleap/fault";
import { getGlobalTestOptions } from "../helpers/global-options-for-tests";
import TrackerSelection from "./TrackerSelection.vue";
import * as rest_querier from "../api/rest-querier";
import type { ProjectInfo, SelectedTracker, TrackerInfo } from "../type";
import type { RetrieveProjects } from "../domain/RetrieveProjects";
import { RetrieveProjectsStub } from "../../tests/stubs/RetrieveProjectsStub";
import { ProjectInfoStub } from "../../tests/stubs/ProjectInfoStub";
import { NOTIFY_FAULT, RETRIEVE_PROJECTS } from "../injection-symbols";

vi.useFakeTimers();

describe("TrackerSelection", () => {
    let errorSpy: Mock,
        projects_retriever: RetrieveProjects,
        first_project: ProjectInfo,
        second_project: ProjectInfo;

    beforeEach(() => {
        errorSpy = vi.fn();
        first_project = ProjectInfoStub.withId(543);
        second_project = ProjectInfoStub.withId(544);
        projects_retriever = RetrieveProjectsStub.withProjects(first_project, second_project);
    });

    function getWrapper(
        selected_trackers: Array<SelectedTracker> = [],
    ): VueWrapper<InstanceType<typeof TrackerSelection>> {
        return shallowMount(TrackerSelection, {
            props: {
                selected_trackers,
            },
            global: {
                ...getGlobalTestOptions(),
                provide: {
                    [NOTIFY_FAULT.valueOf()]: errorSpy,
                    [RETRIEVE_PROJECTS.valueOf()]: projects_retriever,
                },
            },
        });
    }

    describe("mounted()", () => {
        beforeEach(() => {
            vi.spyOn(rest_querier, "getTrackersOfProject").mockReturnValue(okAsync([]));
        });

        it("on init, the projects will be loaded", () => {
            const getSortedProjectsIAmMemberOf = vi.spyOn(
                projects_retriever,
                "getSortedProjectsIAmMemberOf",
            );
            getWrapper();
            expect(getSortedProjectsIAmMemberOf).toHaveBeenCalled();
        });
    });

    describe("loadProjects()", () => {
        beforeEach(() => {
            vi.spyOn(rest_querier, "getTrackersOfProject").mockReturnValue(okAsync([]));
        });

        it("Displays an error when rest route fails", async () => {
            projects_retriever = RetrieveProjectsStub.withFault(Fault.fromMessage("Not Found"));

            const wrapper = getWrapper();
            await vi.runOnlyPendingTimersAsync();

            expect(errorSpy).toHaveBeenCalled();
            expect(errorSpy.mock.calls[0][0].isProjectsRetrieval()).toBe(true);

            expect(wrapper.find("[data-test=tracker-loader]").classes()).not.toContain("fa-spin");
        });

        it("Displays the projects in selectbox", async () => {
            const wrapper = getWrapper();
            await vi.runOnlyPendingTimersAsync();

            expect(wrapper.vm.selected_project).toStrictEqual(first_project);
            expect(wrapper.vm.projects).toStrictEqual([first_project, second_project]);
        });
    });

    describe("loadTrackers()", () => {
        it(`will show a loader and will disable the tracker options that are already selected`, async () => {
            const first_tracker = { id: 8, label: "coquettish" } as TrackerInfo;
            const second_tracker = { id: 26, label: "unfruitfully" } as TrackerInfo;
            const trackers = [first_tracker, second_tracker];
            vi.spyOn(rest_querier, "getTrackersOfProject").mockReturnValue(okAsync(trackers));

            const wrapper = getWrapper([{ tracker_id: 26 } as SelectedTracker]);
            wrapper.vm.selected_project = { id: 102 } as ProjectInfo;
            await nextTick(); // wait for watch
            expect(wrapper.find("[data-test=tracker-loader]").classes()).toContain("fa-spin");

            await vi.runOnlyPendingTimersAsync();
            expect(
                wrapper
                    .findAll("[data-test=cross-tracker-selector-tracker] > option")
                    .map((wrapper) => wrapper.attributes("disabled")),
            ).not.toHaveLength(0);
        });

        it("when there is a REST error, it will be displayed", async () => {
            vi.spyOn(rest_querier, "getTrackersOfProject").mockReturnValue(
                errAsync(Fault.fromMessage("Not Found")),
            );

            getWrapper();
            await vi.runOnlyPendingTimersAsync();

            expect(errorSpy).toHaveBeenCalled();
            expect(errorSpy.mock.calls[0][0].isTrackersRetrieval()).toBe(true);
        });
    });

    describe("addTrackerToSelection()", () => {
        it("when I add a tracker, then an event will be emitted", async () => {
            const tracker = { id: 96, label: "simplus" } as TrackerInfo;
            const tracker_to_add = { id: 97, label: "acinus" } as TrackerInfo;
            vi.spyOn(rest_querier, "getTrackersOfProject").mockReturnValue(
                okAsync([tracker, tracker_to_add]),
            );

            const wrapper = getWrapper();
            await vi.runOnlyPendingTimersAsync(); // load projects
            wrapper.vm.selected_project = second_project;
            await vi.runOnlyPendingTimersAsync(); // load trackers
            wrapper.vm.tracker_to_add = tracker_to_add;
            await nextTick(); // wait for button to be enabled

            await wrapper.get("[data-test=cross-tracker-selector-tracker-button]").trigger("click");

            const emitted = wrapper.emitted("tracker-added");
            if (!emitted) {
                throw Error("Event has not been emitted");
            }
            expect(emitted[0][0]).toStrictEqual({
                selected_project: second_project,
                selected_tracker: tracker_to_add,
            });
            expect(wrapper.vm.tracker_to_add).toBeNull();
        });
    });
});
