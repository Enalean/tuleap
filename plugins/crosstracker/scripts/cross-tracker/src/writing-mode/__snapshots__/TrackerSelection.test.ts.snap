// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerSelection loadProjects() Displays the projects in selectbox 1`] = `
<div
  class="cross-tracker-selector"
>
  <div
    class="tlp-form-element"
    data-test="cross-tracker-selector-global"
  >
    <label
      class="tlp-label"
      for="project"
    >
      
            Project
            
      <i
        aria-hidden="true"
        class="fa-solid fa-asterisk"
      />
    </label>
     
    <select
      class="cross-tracker-selector-project-input tlp-select"
      data-test="cross-tracker-selector-project"
      id="project"
      name="project"
    >
      <option
        value="[object Object]"
      >
        
                unheroically
            
      </option>
      <option
        value="[object Object]"
      >
        
                cycler
            
      </option>
    </select>
  </div>
   
  <div
    class="tlp-form-element tlp-form-element-disabled"
  >
    <label
      class="tlp-label"
      for="tracker"
    >
      
            Tracker
            
      <i
        aria-hidden="true"
        class="fa-solid fa-asterisk"
      />
    </label>
     
    <div
      class="tlp-form-element tlp-form-element-append"
    >
      <select
        class="cross-tracker-selector-tracker-input tlp-select"
        data-test="cross-tracker-selector-tracker"
        disabled="disabled"
        id="tracker"
        name="tracker"
      >
        <option
          class="cross-tracker-please-choose-option"
          value=""
        >
          
                    Please chooseâ€¦
                
        </option>
         
      </select>
       
      <button
        class="tlp-append tlp-button-primary tlp-button-outline"
        data-test="cross-tracker-selector-tracker-button"
        disabled="disabled"
        type="button"
      >
        <i
          aria-hidden="true"
          class="tlp-button-icon fa-solid fa-circle-notch fa-spin"
          data-test="tracker-loader"
        />
        
                Add
            
      </button>
    </div>
  </div>
</div>
`;
